ema.c

```diff
@@ -35,46 +35,49 @@
 #include <assert.h>
 #include <stdlib.h>
 #include <stdio.h>
-#include "ema.h"
+#include "ema_imp.h"
 #include "emalloc.h"
 #include "bit_array.h"
 #include "sgx_mm.h"
 #include "sgx_mm_primitives.h"
 #include "sgx_mm_rt_abstraction.h"
+
 /* State flags */
 #define SGX_EMA_STATE_PENDING   0x8UL
 #define SGX_EMA_STATE_MODIFIED  0x10UL
 #define SGX_EMA_STATE_PR        0x20UL
 #define UNUSED(x) ((void)(x))
-struct ema_t_ {
-    size_t              start_addr;     // starting address, should be on a page boundary
-    size_t              size;           // bytes
-    uint32_t            alloc_flags;    // EMA_RESERVED, EMA_COMMIT_NOW, EMA_COMMIT_ON_DEMAND,
-                                        // OR'ed with EMA_SYSTEM, EMA_GROWSDOWN, ENA_GROWSUP
-    uint64_t            si_flags;       // one of EMA_PROT_NONE, READ, READ_WRITE, READ_EXEC, READ_WRITE_EXEC
-                                        // Or'd with one of EMA_PAGE_TYPE_REG, EMA_PAGE_TYPE_TCS, EMA_PAGE_TYPE_TRIM
-    bit_array *         eaccept_map;    // bitmap for EACCEPT status, bit 0 in eaccept_map[0] for the page at start address
-                                        // bit i in eaccept_map[j] for page at start_address+(i+j<<3)<<12
-    int                 transition;     // state to indicate whether a transition in progress, e.g page type/permission changes.
-    sgx_enclave_fault_handler_t
-                        handler;              // custom PF handler  (for EACCEPTCOPY use)
-    void*               private;        // private data for handler
-    ema_t*              next;           // next in doubly linked list
-    ema_t*              prev;           // prev in doubly linked list
-};
 
+// The head node of the list containing ema nodes belonging to the region [start, end)
 struct ema_root_ {
+    size_t start;      // start address of the region
+    size_t end;        // end address of the region
     ema_t *guard;
 };
-extern size_t mm_user_base;
-extern size_t mm_user_end;
-ema_t dummy_user_ema = {.next = &dummy_user_ema,
-                        .prev = &dummy_user_ema};
-ema_root_t g_user_ema_root = {.guard = &dummy_user_ema};
 
-ema_t dummy_rts_ema = {.next = &dummy_rts_ema,
-                       .prev = &dummy_rts_ema};
-ema_root_t g_rts_ema_root = {.guard = &dummy_rts_ema};
+ema_root_t g_user_ema_root;
+ema_root_t g_rts_ema_root;
+
+ema_t rts_ema_guard  = {.next = &rts_ema_guard, .prev = &rts_ema_guard};
+ema_t user_ema_guard = {.next = &user_ema_guard, .prev = &user_ema_guard};
+
+static void init_ema_root(ema_root_t *root, size_t start, size_t end)
+{
+    root->start = start;
+    root->end = end;
+}
+
+void init_rts_ema_root(size_t start, size_t end)
+{
```

Comment:
Can we remove this API for RTS? Spec for sgx_mm_init only requires user range specified. All address outside user range can be used by RTS.